cpp/src/arrow/filesystem/gcsfs_internal.cc

```diff
@@ -62,6 +65,133 @@ Status ToArrowStatus(const google::cloud::Status& s) {
+    const std::shared_ptr<const KeyValueMetadata>& metadata) {
+  if (!metadata) {
+    return gcs::PredefinedAcl{};
+  }
+
+  const auto& keys = metadata->keys();
+  const auto& values = metadata->values();
+
+  for (std::size_t i = 0; i < keys.size(); ++i) {
+    if (keys[i] == "predefinedAcl") {
+      return gcs::PredefinedAcl(values[i]);
+    }
+  }
+  return gcs::PredefinedAcl{};
+}
+
+Result<gcs::WithObjectMetadata> ToObjectMetadata(
+    const std::shared_ptr<const KeyValueMetadata>& metadata) {
+  if (!metadata) {
+    return gcs::WithObjectMetadata{};
+  }
+
+  static auto const setters = [] {
+    using setter = std::function<Status(gcs::ObjectMetadata&, const std::string&)>;
+    return std::map<std::string, setter>{
+        {"cacheControl",
+         [](gcs::ObjectMetadata& m, const std::string& v) {
+           m.set_cache_control(v);
+           return Status::OK();
+         }},
+        {"contentDisposition",
+         [](gcs::ObjectMetadata& m, const std::string& v) {
+           m.set_content_disposition(v);
+           return Status::OK();
+         }},
+        {"contentEncoding",
+         [](gcs::ObjectMetadata& m, const std::string& v) {
+           m.set_content_encoding(v);
+           return Status::OK();
+         }},
+        {"contentLanguage",
+         [](gcs::ObjectMetadata& m, const std::string& v) {
+           m.set_content_language(v);
+           return Status::OK();
+         }},
+        {"contentType",
+         [](gcs::ObjectMetadata& m, const std::string& v) {
+           m.set_content_type(v);
+           return Status::OK();
+         }},
+        {"customTime",
+         [](gcs::ObjectMetadata& m, const std::string& v) {
+           std::string err;
+           absl::Time t;
+           if (!absl::ParseTime(absl::RFC3339_full, v, &t, &err)) {
+             return Status::Invalid("Error parsing RFC-3339 timestamp: '", v, "': ", err);
+           }
+           m.set_custom_time(absl::ToChronoTime(t));
+           return Status::OK();
+         }},
+        {"storageClass",
+         [](gcs::ObjectMetadata& m, const std::string& v) {
+           m.set_storage_class(v);
+           return Status::OK();
+         }},
+        {"predefinedAcl",
+         [](gcs::ObjectMetadata&, const std::string&) { return Status::OK(); }},
+        {"encryptionKeyBase64",
+         [](gcs::ObjectMetadata&, const std::string&) { return Status::OK(); }},
+        {"kmsKeyName",
+         [](gcs::ObjectMetadata&, const std::string&) { return Status::OK(); }},
+    };
+  }();
+
+  const auto& keys = metadata->keys();
+  const auto& values = metadata->values();
+
+  gcs::ObjectMetadata object_metadata;
+  for (std::size_t i = 0; i < keys.size(); ++i) {
+    auto it = setters.find(keys[i]);
+    if (it != setters.end()) {
+      auto status = it->second(object_metadata, values[i]);
+      if (!status.ok()) return status;
+    } else {
+      object_metadata.upsert_metadata(keys[i], values[i]);
```

Comment:
This is meant to allow inserting arbitrary metadata strings? Will GCS balk if the user throws some unrecognized metadata keys here?

Reply:
GCS accepts (mostly) arbitrary metadata keys:

https://cloud.google.com/storage/docs/metadata#custom-metadata